metadata/buckets.go

```diff
@@ -182,6 +185,39 @@ func createBucketIfNotExists(tx *bolt.Tx, keys ...[]byte) (*bolt.Bucket, error)
 	return bkt, nil
 }
 
+func namespacesBucketPath() []byte {
+	return bucketKeyVersion
+}
+
+func getNamespacesBucket(tx *bolt.Tx) *bolt.Bucket {
+	return getBucket(tx, namespacesBucketPath())
+}
+
+func getSharableBucket(tx *bolt.Tx, dgst digest.Digest) *bolt.Bucket {
+	var bkt *bolt.Bucket
+	nsbkt := getNamespacesBucket(tx)
+	cur := nsbkt.Cursor()
+	for k, _ := cur.First(); k != nil; k, _ = cur.Next() {
```

Comment:
To reduce the search time, what about having an auxiliary bucket that only holds shared namespaces labels. When it is created, add it to both and use the auxiliary one to search.  Of course it will bring in additional work of adding deleting to/from 2 buckets.